File: datastructures/linked-list/singly_linked_list.py
# Node Class
class Node:
    def __init__(self, data):
        self.data = data
        self.next = None

# List Class
class UnorderedList:
    def __init__(self):
        self.head = None

    # returns true if empty
    def isEmpty(self):
        return self.head == None

    # adds new element to the front of the list
    def add(self, item):
        temp = Node(item)
        temp.next = self.head
        self.head = temp

    # returns dynamically counted size
    def size(self):
        current = self.head
        count = 0
        while current != None:
            count += 1
            current = current.next

        return count

    # returns node position
    def search(self, item):
        current = self.head
        count = 0
        while current != None:
            if current.data == item:
                return count
            count += 1
            current = current.next
        return -1

    # returns true if the element is in the list
    def exists(self, item):
        return self.search(item) != -1

    # removes first occurence of the item
    def remove(self, item):
        current = self.head
        previous = None
        found = False
        while not found and current != None:
            if current.data == item:
                found = True
            else:
                previous = current
                current = current.next

        if current != None:
            if previous == None:
                self.head = current.next
            else:
                previous.next = current.next

    # visualizes the list for print function
    def __str__(self):
        current = self.head
        visualized = ''
        while current != None:
            visualized += '{} '.format(current.data)
            if current.next != None: visualized += '-> '
            current = current.next

        return visualized


if __name__ == '__main__':
    dummyList = UnorderedList()
    dummyList.add(4)
    dummyList.add(3)
    dummyList.add(2)
    dummyList.add(6)
    dummyList.add(7)

    # Print the list
    print(dummyList)

    # size
    print(dummyList.size())

    # isEmpty
    print(dummyList.isEmpty())

    # search
    print(dummyList.search(2))
    print(dummyList.search(9))

    # exists
    print(dummyList.exists(2))
    print(dummyList.exists(9))

    # remove
    dummyList.remove(2)
    dummyList.remove(9)

    print(dummyList)

File: datastructures/linked-list/singly_linked_list.cc
#include <iostream>
#include <string>
#include <stdexcept>

// Node Structure
struct node {
    int data;
    node *next;

    // Default constructor - resets data to 0.
    node() {
        data = 0;
        next = NULL;
    }

    // Set data constructor
    node(int data) {
        this->data = data;
        next = NULL;
    }

    // Set data and next constructor
    node(int data, node *next) {
        this->data = data;
        this->next = next;
    }
};

class linked_list {
    node *head; // Head pointer

public:
    // Default constructor
    linked_list() {
        head = new node();
    }

    // Set head value constructor
    linked_list(int data) {
        head = new node(data);
    }

    // Set head node constructor
    linked_list(node *head) {
        this->head = head;
    }

    // check if the list is empty
    bool is_empty() {
        return !head;
    }

    // push value
    node* push(int data) {
        node *temp = new node(data);
        temp->next = head;
        return head = temp;
    }

    // push node
    node* push(node *new_node) {
        new_node->next = head;
        return head = new_node;
    }

    // pop node
    int pop() {
        if (this->is_empty()) {
            throw "Empty List Exception";
        }
        node *temp = head;
        int val = temp->data;
        head = head->next;
        delete temp;
        return val;
    }

    // peek head node value
    int peek() {
        if (this->is_empty()) {
            throw "Empty List Exception";
        }
        return head->data;
    }

    // insert data at particular position
    node* insert(int data, int position) {
        return this->insert(new node(data), position);
    }

    // insert node at particular position
    node* insert(node *new_node, int position) {
        if (this->is_empty()) {
            return head = new_node;
        }

        node *current = head;
        position--;
        while (position && current->next) {
            current = current->next;
            position--;
        }
       new_node->next = current->next;
       current->next = new_node;

       return head;
    }

    // remove first occurence of data
    void remove(int data) {
        if (this->is_empty()) {
            throw "Empty List Exception";
        }

        if (head->data == data) {
            this->pop();
        }

        node *current = head;
        while (current->next && current->next->data != data) {
            current = current->next;
        }

        if (current->next) {
            node *temp = current->next;
            current->next = temp->next;
            delete temp;
        }
    }

    // Visualize the list
    std::string show() {
        node *current = head;
        std::string visualized_list = "";
        if (this->is_empty()) {
            return "";
        }

        if (current && !current->next) {
            return std::to_string(current->data);
        }

        while (current->next) {
            visualized_list += std::to_string(current->data) + " -> ";
            current = current->next;
        }
        return visualized_list.substr(0, visualized_list.length() - 4);
    }

    // destructor
    ~linked_list() {
        node *current;
        while (head) {
            current = head;
            head = head->next;
            delete current;
        }
    }
};

int main(int argc, char *argv[]) {
    linked_list my_list;
    my_list.push(3);
    my_list.push(4);
    my_list.push(5);
    my_list.push(6);
    my_list.push(7);
    my_list.push(8);
    my_list.push(9);
    my_list.push(10);
    my_list.push(11);
    std::cout << my_list.show() << "\n";

    my_list.pop();
    my_list.pop();
    my_list.pop();
    my_list.pop();
    std::cout << my_list.show() << "\n";

    my_list.insert(15, 3);
    my_list.remove(6);
    my_list.remove(8);
    std::cout << my_list.show() << "\n";
}
